src/optimise1.jl

```diff
@@ -0,0 +1,248 @@
+using IntervalRootFinding, IntervalArithmetic, StaticArrays, ForwardDiff, BenchmarkTools, Compat, IntervalOptimisation, DataStructures, IntervalConstraintProgramming
+
+import Base.isless
+
+struct IntervalMinima{T<:Real}
+    intval::Interval{T}
+    global_minimum::T
+end
+
+function isless(a::IntervalMinima{T}, b::IntervalMinima{T}) where {T<:Real}
+    return isless(a.global_minimum, b.global_minimum)
+end
+
+function minimise1d(f::Function, x::Interval{T}; reltol=1e-3, abstol=1e-3) where {T<:Real}
+
+    Q = binary_minheap(IntervalMinima{T})
+
+    global_minimum = f(interval(mid(x))).hi
+
+    arg_minima = Interval{T}[]
+
+    push!(Q, IntervalMinima(x, global_minimum))
+
+    while !isempty(Q)
+
+        p = pop!(Q)
+
+        if isempty(p.intval)
+            continue
+        end
+
+        if p.global_minimum > global_minimum
+            continue
+        end
+
+        current_minimum = f(interval(mid(p.intval))).hi
+
+        if current_minimum < global_minimum
+            global_minimum = current_minimum
+        end
+
+        if diam(p.intval) < abstol
+            push!(arg_minima, p.intval)
+        else
+            x1, x2 = bisect(p.intval)
+            push!(Q, IntervalMinima(x1, f(x1).lo), IntervalMinima(x2, f(x2).lo))
+        end
+    end
+    lb = minimum(inf.(f.(arg_minima)))
+
+    return lb..global_minimum, arg_minima
+end
+
+function minimise1d_deriv(f::Function, x::Interval{T}; reltol=1e-3, abstol=1e-3) where {T<:Real}
+
+    Q = binary_minheap(IntervalMinima{T})
+
+    global_minimum = f(interval(mid(x))).hi
+
+    arg_minima = Interval{T}[]
+
+    push!(Q, IntervalMinima(x, global_minimum))
+
+    while !isempty(Q)
+
+        p = pop!(Q)
+
+        if isempty(p.intval)
+            continue
+        end
+
+        if p.global_minimum > global_minimum
+            continue
+        end
+
+        deriv = ForwardDiff.derivative(f, p.intval)
+        if 0 ∉ deriv
+            continue
+        end
+        # Second derivative contractor
+        # doublederiv = ForwardDiff.derivative(x->ForwardDiff.derivative(f, x), p.intval)
+        # if doublederiv < 0
+        #     continue
+        # end
+
+        m = mid(p.intval)
+
+        current_minimum = f(interval(m)).hi
+
+        if current_minimum < global_minimum
+            global_minimum = current_minimum
+        end
+
+
+        # Contractor 1
+        x = m .+ extended_div((interval(-∞, global_minimum) - f(m)), deriv)
```

Comment:
What is this for?

Reply:
This is the derivative contractor as given in the Jaulin book